include/exadg/acoustic_conservation_equations/spatial_discretization/operators/grad_div_operator.h

```diff
@@ -0,0 +1,479 @@
+  {
+    do_evaluate(dst, src, false, time);
+  }
+
+private:
+  void
+  do_evaluate(BlockVectorType &       dst,
+              BlockVectorType const & src,
+              double const            time,
+              bool const              zero_dst_vector) const
+  {
+    if(data.formulation == FormulationGradientDivergenceTerms::Strong)
+    {
+      do_evaluate_formulation<FormulationGradientDivergenceTerms::Strong>(dst,
+                                                                          src,
+                                                                          time,
+                                                                          zero_dst_vector);
+    }
+    else if(data.formulation == FormulationGradientDivergenceTerms::Weak)
+    {
+      do_evaluate_formulation<FormulationGradientDivergenceTerms::Weak>(dst,
+                                                                        src,
+                                                                        time,
+                                                                        zero_dst_vector);
+    }
+    else
+    {
+      AssertThrow(false, dealii::ExcMessage("Not implemented."));
+    }
+  }
+
+  template<FormulationGradientDivergenceTerms formulation>
+  void
+  do_evaluate_formulation(BlockVectorType &       dst,
+                          BlockVectorType const & src,
+                          double const            time,
+                          bool const              zero_dst_vector) const
+  {
+    evaluation_time = (Number)time;
+
+    matrix_free->loop(&This::cell_loop<formulation>,
+                      &This::face_loop<formulation>,
+                      &This::boundary_face_loop<formulation>,
+                      this,
+                      dst,
+                      src,
+                      zero_dst_vector,
+                      dealii::MatrixFree<dim, Number>::DataAccessOnFaces::values,
+                      dealii::MatrixFree<dim, Number>::DataAccessOnFaces::values);
+  }
+
+  template<FormulationGradientDivergenceTerms formulation>
+  void
+  do_cell_integral(CellIntegratorP & pressure,
+                   CellIntegratorU & velocity,
+                   Number const      rho_inv,
+                   Number const      rhocc) const
+  {
+    for(unsigned int q : pressure.quadrature_point_indices())
+    {
+      if constexpr(formulation == FormulationGradientDivergenceTerms::Weak)
+      {
+        pressure.submit_gradient(rhocc * kernel.get_volume_flux_weak_momentum(velocity, q), q);
+        velocity.submit_divergence(rho_inv * kernel.get_volume_flux_weak_mass(pressure, q), q);
+      }
+      else if constexpr(formulation == FormulationGradientDivergenceTerms::Strong)
+      {
+        pressure.submit_value(rhocc * kernel.get_volume_flux_strong_momentum(velocity, q), q);
+        velocity.submit_value(rho_inv * kernel.get_volume_flux_strong_mass(pressure, q), q);
+      }
```

Comment:
If you care about computational efficiency, I would suggest a third alternative, namely to use the weak form `-(grad q, rho u)` and `(w, grad p)` with the respective face terms, see also https://github.com/kronbichler/exwave/blob/97000e8234f41c6cdd7ccd46ffb37979dd922644/source/wave_equation_operations.cc#L447-L451 where we did this based on a large number of efficiency experiments. The important ingredient is to use the gradient only on the scalar variable. This can be explained by the underlying sum factorization algorithms: If we are in 3D and compute the gradient of a scalar variable, we need 6 sum factorization sweeps (meaning 1D interpolation kernels going through all points), 3 for transforming each spatial direction to quadrature points and 3 for the components of the gradient in collocation space. The divergence of a vector field needs 18, 6 for each component. One could suspect that the evaluation routine could know whether we only need the divergence and the mesh could possibly be Cartesian, where in fact 9 sweeps would suffice, but we don't know that. The value of a vector-valued variable needs 9 sum factorization sweeps. The total cost is then `(6 + 9) * 2 = 30` sweeps for both the pressure and velocity terms. In your formulation you need `(3 + 18) + (6 + 9) = 36` sweeps, where the two terms are the first and second line of the equation in the PR description, respectively.

As a second comment, I note that your code is dangerous in the sense that `pressure.submit_gradient()` gets called before the `pressure` field is handed into the volume mass flux. Since the `get_xxx()` and `submit_xxx` refer to the same underlying storage, it is very easy to make the mistake we the `get_volume_flux_weak_mass` would also request the `pressure.get_gradient()` and then produce wrong results. Your code is fine because you request the value of the pressure (after you submit the gradient), which are in fact different places in memory. I think brittle behavior of `CellIntegrator` is something we need to think about in a future design update. (It is very important that those refer to the same data field when seen for all quadrature points because otherwise the caches would be overwhelmed, but while sitting on an individual point another interface would be better.) If you would switch to the code suggested by me, you would have to take this into account, because then we would always step on another function at some place.

As a final remark, the form suggested by me should also be the most robust one, at least according to my understanding of the wave equation. It leads to a skew-symmetric formulation, so even if your numerical quadrature is not exact (non-affine cells), the energy will be non-increasing. In either the `Weak` or `Strong` form derived by you, this cannot be mathematically guaranteed, even though it will often be OK.

Reply:
Thank you for the detailed review and the helpful suggestions. 

To avoid any potential future problems with the same underlying memory that is used by `evaluate()` and `submit()` I cache the fluxes using `get_value()` before submitting anything to the integrators. [7ed38c1](https://github.com/exadg/exadg/pull/587/commits/7ed38c127cf514012d8638793b320a4fd8403453)

I added the proposed formulation in [cff4e25](https://github.com/exadg/exadg/pull/587/commits/cff4e25dfca35317d155859f278d1658ad79167b) and made it the default one due to all the arguments you made.

I agree that the skew-symmetric formulation is the best choice. Quite some time ago I tested all formulations (including the skew-symmetric one) and could not see any differences in the results in practice. I didn't think about the performance implications back then, and obviously it is better to have the skew-symmetric formulation either way, since all the mentioned advantages come without drawbacks.

Reply:
Yes, it is good to have all options available. Since part of our work in ExaDG is developing discretization methods, it is important to be able to switch between different variants at any time by just changing some input parameters.